x/precisebank/keeper/fractional_balance.go

```diff
@@ -0,0 +1,104 @@
+package keeper
+
+import (
+	"fmt"
+
+	sdkmath "cosmossdk.io/math"
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/kava-labs/kava/x/precisebank/types"
+)
+
+// GetFractionalBalance returns the fractional balance for an address.
+func (k *Keeper) GetFractionalBalance(
+	ctx sdk.Context,
+	address sdk.AccAddress,
+) (sdkmath.Int, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.FractionalBalancePrefix)
+
+	bz := store.Get(types.FractionalBalanceKey(address))
+	if bz == nil {
+		return sdkmath.ZeroInt(), false
+	}
+
+	var bal sdkmath.Int
+	if err := bal.Unmarshal(bz); err != nil {
+		panic(fmt.Errorf("failed to unmarshal fractional balance: %w", err))
+	}
+
+	return bal, true
+}
+
+// SetFractionalBalance sets the fractional balance for an address.
+func (k *Keeper) SetFractionalBalance(
+	ctx sdk.Context,
+	address sdk.AccAddress,
+	amount sdkmath.Int,
+) {
+	if amount.IsZero() {
+		k.DeleteFractionalBalance(ctx, address)
+		return
+	}
+
+	// Ensure the fractional balance is valid before setting it. Use the
+	// NewFractionalAmountFromInt wrapper to use its Validate() method.
+	if err := types.NewFractionalAmountFromInt(amount).Validate(); err != nil {
+		panic(fmt.Errorf("amount is invalid: %w", err))
```

Comment:
yeah I was thinking about this too, my thought was these store methods are used internally and receiving an invalid value is unrecoverable and incorrect behavior, like if something went wrong with Marshalling, something went real wrong.

But I'm also not entirely sure if that's consistent or ideal since it receiving an invalid value would be a concern of something external from the method, so yup it would be good to define something more clear for us to be consistent on when to panic and when to error 😃